internal/component/component.go

```diff
@@ -0,0 +1,153 @@
+// Package component provides types for representing discovered Terragrunt components.
+//
+// These include units and stacks.
+//
+// This package contains only data types and their associated methods, with no discovery logic.
+// It exists separately from the discovery package to allow other packages (like filter) to
+// depend on these types without creating circular dependencies.
+package component
+
+import (
+	"sort"
+
+	"github.com/gruntwork-io/terragrunt/config"
+	"github.com/gruntwork-io/terragrunt/internal/errors"
+)
+
+const (
+	// Unit is a component of kind unit.
+	Unit Kind = "unit"
+	// Stack is a component of kind stack.
+	Stack Kind = "stack"
+)
+
+// Kind is the type of Terragrunt component.
+type Kind string
+
+// Component represents a discovered Terragrunt configuration.
+type Component struct {
+	Parsed           *config.TerragruntConfig
+	DiscoveryContext *DiscoveryContext
+
+	Kind Kind
```

Comment:
We may need a "version" field to support multiple implementations 🤔

Reply:
We can add a top-level version field in each HCL file to support future syntax or behavior changes without breaking existing setups.